std/regex/package.d

```diff
@@ -821,43 +821,64 @@ public:
     @property inout(Captures!R) captures() inout { return _captures; }
 }
 
-private @safe auto matchOnce(RegEx, R)(R input, const auto ref RegEx prog) pure
+private auto matchOnceImpl(RegEx, R)(R input, const auto ref RegEx prog) @trusted
 {
-    auto matchOnceImp = () @trusted {
-        alias Char = BasicElementOf!R;
-        static struct Key
-        {
-            immutable(Char)[] pattern;
-            uint flags;
-        }
-        static Key cacheKey = Key("", -1);
-        static Matcher!Char cache;
-        auto factory = prog.factory is null ? defaultFactory!Char(prog) : prog.factory;
-        auto key = Key(prog.pattern, prog.flags);
-        Matcher!Char engine;
-        if (cacheKey == key)
-        {
-            engine = cache;
-            engine.rearm(input);
-        }
-        else
-        {
-            engine = factory.create(prog, input);
-            if (cache) factory.decRef(cache); // destroy cached engine *after* building a new one
-            cache = engine;
-            cacheKey = key;
-        }
-        auto captures = Captures!R(input, prog.ngroup, prog.dict);
-        captures.matches.mutate((slice) pure { captures._nMatch = engine.match(slice); });
-        return captures;
-    };
-
-    // this should be faked as pure because the static mutable variables are
-    // used to cache the key and the character matcher
-    alias T = typeof(matchOnceImp);
+    alias Char = BasicElementOf!R;
+    static struct Key
+    {
+        immutable(Char)[] pattern;
+        uint flags;
+    }
+    static Key cacheKey = Key("", -1);
+    static Matcher!Char cache;
+    auto factory = prog.factory is null ? defaultFactory!Char(prog) : prog.factory;
+    auto key = Key(prog.pattern, prog.flags);
+    Matcher!Char engine;
+    if (cacheKey == key)
+    {
+        engine = cache;
+        engine.rearm(input);
+    }
+    else
+    {
+        engine = factory.create(prog, input);
+        if (cache) factory.decRef(cache); // destroy cached engine *after* building a new one
+        cache = engine;
+        cacheKey = key;
+    }
+    auto captures = Captures!R(input, prog.ngroup, prog.dict);
+    captures.matches.mutate((slice) pure { captures._nMatch = engine.match(slice); });
+    return captures;
+}
+
+// Used to generate a pure wrapper for matchOnceImpl. Based on the example in the
+// std.traits.SetFunctionAttributes documentation.
+private auto assumePureFunction(T)(T t)
```

Comment:
Are we worried about this function template seemingly being duplicated in the other file?

Reply:
Where would you put it with public visibility ? these kind of functions are hacks so this is okay.
Out of topic but eventually D should allow things like  

```d
void bar(){}
void foo() pure
{
     (cast(pure) bar)();
}
```

instead of using these assumeNothrow assumeNoGc etc. templates